processing/src/main/java/io/druid/segment/virtual/SingleStringInputDimensionSelector.java

```diff
@@ -0,0 +1,156 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.segment.virtual;
+
+import com.google.common.base.Preconditions;
+import com.google.common.base.Predicate;
+import io.druid.java.util.common.ISE;
+import io.druid.math.expr.Expr;
+import io.druid.math.expr.Parser;
+import io.druid.query.filter.ValueMatcher;
+import io.druid.query.monomorphicprocessing.RuntimeShapeInspector;
+import io.druid.segment.DimensionSelector;
+import io.druid.segment.DimensionSelectorUtils;
+import io.druid.segment.IdLookup;
+import io.druid.segment.data.IndexedInts;
+import io.druid.segment.data.SingleIndexedInt;
+
+import javax.annotation.Nullable;
+
+/**
+ * A DimensionSelector decorator that computes an expression on top of it.
+ */
+public class SingleStringInputDimensionSelector implements DimensionSelector
+{
+  private final DimensionSelector selector;
+  private final Expr expression;
+  private final SingleInputBindings bindings = new SingleInputBindings();
+
+  // 0 if selector has null as a value; 1 if it doesn't.
+  private final int nullAdjustment;
+
+  public SingleStringInputDimensionSelector(
+      final DimensionSelector selector,
+      final Expr expression
+  )
+  {
+    // Verify expression has just one binding.
+    if (Parser.findRequiredBindings(expression).size() != 1) {
+      throw new ISE("WTF?! Expected expression with just one binding");
+    }
+
+    // Verify selector has a working dictionary.
+    if (selector.getValueCardinality() == DimensionSelector.CARDINALITY_UNKNOWN
+        || !selector.nameLookupPossibleInAdvance()) {
+      throw new ISE("Selector of class[%s] does not have a dictionary, cannot use it.", selector.getClass().getName());
+    }
+
+    this.selector = Preconditions.checkNotNull(selector, "selector");
+    this.expression = Preconditions.checkNotNull(expression, "expression");
+    this.nullAdjustment = selector.getValueCardinality() == 0 || selector.lookupName(0) != null ? 1 : 0;
+  }
+
+  @Override
+  public void inspectRuntimeShape(final RuntimeShapeInspector inspector)
+  {
+    inspector.visit("selector", selector);
+    inspector.visit("expression", expression);
+  }
+
+  @Override
+  public IndexedInts getRow()
+  {
+    // Treat any non-single-valued row as a row containing a single null value, to ensure consistency with
+    // other expression selectors. See also ExpressionSelectors.supplierFromDimensionSelector for similar behavior.
+    final IndexedInts row = selector.getRow();
+
+    if (row.size() == 1) {
+      if (nullAdjustment == 0) {
+        return row;
+      } else {
+        return new SingleIndexedInt(row.get(0) + nullAdjustment);
```

Comment:
Could add `MutableSingleIndexedInt` and cache the object

Reply:
I'm not sure if this offers a measurable benefit and I'm also not sure if the return value from `getRow` ever needs to persist beyond the next movement of the selector. So I left this as-is.

Reply:
ok